client/state/data-layer/wpcom-http/utils.js

```diff
@@ -161,3 +166,101 @@ export const dispatchRequest = ( initiator, onSuccess, onError, options = {} ) =
 
 	return initiator( store, action );
 };
+
+/**
+ * Dispatches to appropriate function based on HTTP request meta
+ *
+ * @see state/data-layer/wpcom-http/actions#fetch creates HTTP requests
+ *
+ * When the WPCOM HTTP data layer handles requests it will add
+ * response data and errors to a meta property on the given success
+ * error, and progress handling actions.
+ *
+ * This function accepts several functions as the fetch, success, error and
+ * progress handlers for actions and it will call the appropriate
+ * one based on the stored meta.
+ *
+ * These handlers are action creators: based on the input data coming from the HTTP request,
+ * it will return an action (or an action thunk) to be executed as a response to the given
+ * HTTP event.
+ *
+ * If both error and response data is available this will call the
+ * error handler in preference over the success handler, but the
+ * response data will also still be available through the action meta.
+ *
+ * The functions should conform to the following type signatures:
+ *   fetch  :: Action -> Action (action creator with one Action parameter)
+ *   onSuccess  :: Action -> ResponseData -> Action (action creator with two params)
+ *   onError    :: Action -> ErrorData -> Action
+ *   onProgress :: Action -> ProgressData -> Action
+ *   fromApi    :: ResponseData -> TransformedData throws TransformerError|SchemaError
+ *
+ * @param {Object} options object with named parameters:
+ * @param {Function} fetch called if action lacks response meta; should create HTTP request
+ * @param {Function} onSuccess called if the action meta includes response data
+ * @param {Function} onError called if the action meta includes error data
+ * @param {Function} onProgress called on progress events when uploading
+ * @param {Function} fromApi maps between API data and Calypso data
+ * @returns {Action} action or action thunk to be executed in response to HTTP event
+ */
+export const dispatchRequestEx = options => {
+	if ( ! options.fetch ) {
+		warn( 'fetch handler is not defined: no request will ever be issued' );
+	}
+
+	if ( ! options.onSuccess ) {
+		warn( 'onSuccess handler is not defined: response to the request is being ignored' );
+	}
+
+	if ( ! options.onError ) {
+		warn( 'onError handler is not defined: error during the request is being ignored' );
+	}
+
+	return ( store, action ) => {
+		// create the low-level action we want to dispatch
+		const requestAction = createRequestAction( options, action );
+		// dispatch the low level action (if any was created) and return the result
+		return requestAction ? store.dispatch( requestAction ) : undefined;
```

Comment:
Not a big deal here, but if we change the return value from `createRequestAction` to a list we can just dispatch each item and then the empty case will be handled automatically by means of the safety that empty lists already provide.

```js
processAction( options, action ).forEach( store.dispatch );
```

Reply:
`dispatch` can return a value and I'd like to avoid adding places where it's thrown away. I have some ideas down the road that could greatly improve data layer ergonomics and these return values will be important there.

Using the array is a nice monadic trick to avoid a conditional, but I think it's an unnecessary complication -- you need to stop and think for a while to decode what it does.

Reply:
I'm just thinking of cases where we know we have arrays to return, plus the notion that the data layer doesn't return anything - it's all loosely coupled asynchronously

did you have something you know you want to return from dispatch and use somewhere?

Reply:
I have many diverse use cases where returning a Promise from the dispatch would make complicated things simple and impossible things possible. Without violating any of the data layer design principles. But that's definitely for another PR -- we will surely have a vigorous debate about that soon 😉